### PR TITLE
[SPARK-16991][SPARK-17099][SPARK-17120][SQL] Fix Outer Join Elimination when Filter's isNotNull Constraints Unable to Filter Out All Null-supplying Rows

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?

This PR is to fix an incorrect outer join elimination when filter's `isNotNull` constraints is unable to filter out all null-supplying rows. For example, `isnotnull(coalesce(b#227, c#238))`.

Users can hit this error when they try to use `using/natural outer join`, which is converted to a normal outer join with a `coalesce` expression on the `using columns`. For example, 

``` Scala
    val a = Seq((1, 2), (2, 3)).toDF("a", "b")
    val b = Seq((2, 5), (3, 4)).toDF("a", "c")
    val c = Seq((3, 1)).toDF("a", "d")
    val ab = a.join(b, Seq("a"), "fullouter")
    ab.join(c, "a").explain(true)
```

The dataframe `ab` is doing `using full-outer join`, which is converted to a normal outer join with a `coalesce` expression. Constraints inference generates a `Filter` with constraints `isnotnull(coalesce(b#227, c#238))`. Then, it triggers a wrong outer join elimination and generates a wrong result.

```
Project [a#251, b#227, c#237, d#247]
+- Join Inner, (a#251 = a#246)
   :- Project [coalesce(a#226, a#236) AS a#251, b#227, c#237]
   :  +- Join FullOuter, (a#226 = a#236)
   :     :- Project [_1#223 AS a#226, _2#224 AS b#227]
   :     :  +- LocalRelation [_1#223, _2#224]
   :     +- Project [_1#233 AS a#236, _2#234 AS c#237]
   :        +- LocalRelation [_1#233, _2#234]
   +- Project [_1#243 AS a#246, _2#244 AS d#247]
      +- LocalRelation [_1#243, _2#244]

== Optimized Logical Plan ==
Project [a#251, b#227, c#237, d#247]
+- Join Inner, (a#251 = a#246)
   :- Project [coalesce(a#226, a#236) AS a#251, b#227, c#237]
   :  +- Filter isnotnull(coalesce(a#226, a#236))
   :     +- Join FullOuter, (a#226 = a#236)
   :        :- LocalRelation [a#226, b#227]
   :        +- LocalRelation [a#236, c#237]
   +- LocalRelation [a#246, d#247]
```

**A note to the `Committer`**, please also give the credit to @dongjoon-hyun who submitted another PR for fixing this issue. https://github.com/apache/spark/pull/14580
### How was this patch tested?

Added test cases

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @hvanhovell @yhuai @dongjoon-hyun @nsyca 

also cc the original reviewers: @davies @sameeragarwal @marmbrus 

Thanks!

**Comment by nsyca** [comment]
This code change looks good to me. The logic in

``
filter.constraints.filter(_.isInstanceOf[IsNotNull])
      .exists(expr => join.left.outputSet.intersect(expr.references).nonEmpty)
``
is useless here. Checking just an ISNOTNULL() predicate that has an argument from the left table is not sufficient to make the conclusion that the predicate will filter null rows from the left table. Likewise for the same logic checking on the right table.

One minor comment: Does filter.constraints subsume filter.condition? By the reading of the lazy evaluation of the ``val constraints`` in ``QueryPlan.scala`` and the ``override protected def validConstraints`` in ``case class Filter``, I think it does. If so, then the ``++`` in the ``val conditions`` at line 1331 is redundant. It could just be changed to ``val conditions = filter.constraints``.

I will not monitor this PR until my return on 8/29.

**Comment by rxin** [comment]
cc @sameeragarwal

**Comment by gatorsmile** [comment]
@nsyca `filter.constraints` is not a superset of `filter.conditions`. We are doing the same in the rule `InferFiltersFromConstraints`: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala#L756

**Comment by hvanhovell** [comment]
@gatorsmile could you add cases for both [SPARK-17099](https://issues.apache.org/jira/browse/SPARK-17099) and [SPARK-17120](https://issues.apache.org/jira/browse/SPARK-17120)?

**Comment by hvanhovell** [comment]
LGTM - pending test cases for SPARK-17099 and SPARK-17120.

**Comment by gatorsmile** [comment]
Will do it soon. Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 1]
```diff
@@ -2586,6 +2586,63 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
...
 
+  test("SPARK-16991: Full outer join followed by inner join produces wrong results") {
+    val a = Seq((1, 2), (2, 3)).toDF("a", "b")
+    val b = Seq((2, 5), (3, 4)).toDF("a", "c")
+    val c = Seq((3, 1)).toDF("a", "d")
+    val ab = a.join(b, Seq("a"), "fullouter")
+    checkAnswer(ab.join(c, "a"), Row(3, null, 4, 1) :: Nil)
+  }
+
+  test("SPARK-17099: Incorrect result when HAVING clause is added to group by query") {
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: move these over to SQLQueryTestSuite?

Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: I thought Spark 2.0.1 did not have `SQLQueryTestSuite`. Let me do it now. Thanks!

Comment (hvanhovell) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: We backported it.

**Comment by gatorsmile** [comment]
@rxin @hvanhovell Moved the test cases. Just let me know if any change is needed. Thanks!

**Comment by hvanhovell** [comment]
LGTM - merging to master/2.0. Thanks!

**Pull request closed without merging**